Remove all instructions in a block after OpKill.

perlfu opened the pull request:
During inlining it is possible to construct a block which has trailing instructions after a kill.
This can prevent return block unification and cause exports to be skipped.
In particular some SPIRV generators terminate a block with a function calling OpKill with OpUnreachable, this also fixes these by replacing unreachable with ret void, which is then unified.

amdvlk-admin commented:
### Test summary for commit 230d3105cf62ca8ba352b037c0fb961ce2ad959b
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev c9937b277c491a55c689eec7cdd48fb238b8004c</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev c6c90450fb1abf5f414acf1e38a0f51a72c426c1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1598ada40008302abedfd724275329f8e7dea637</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24006/46960 (51.1%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 22954/46960 (48.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24005/46960 (51.1%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 22955/46960 (48.9%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22647/46939 (48.2%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 24292/46939 (51.8%)</li>
</ul></details>


</ul></details>

kuhar reviewed (commented):
Can there be some other instructions between `kill` and `unreachable`?

perlfu commented:
> Can there be some other instructions between `kill` and `unreachable`?

In the SPIRV there should be nothing after OpKill as it is a terminator; however, in expansion we turn OpKill into kill followed by a return.  The sequence of kill; unreachable; occurs when a kill function that ends in a kill is inline in a block that is just call; unreachable.  This is demonstrated in the test in this PR.
In principle I guess we could have a sequence of call; call; unreachable; which becomes kill; some code; unreachable.
So in practice it might be better to remove all instructions in a block after a kill and replace these with a branch?

perlfu commented:
I have revised this to more generally remove all instructions inlined after a kill.

kuhar commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,138 @@
+  visit(m_module);
+
+  // Remove any dead instructions.
+  while (m_removalStack.size() > 0) {
```
tiny nit: `while (!m_removalStack.empty())`

kuhar commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,138 @@
+
+  // Remove any dead instructions.
+  while (m_removalStack.size() > 0) {
+    auto deadInst = m_removalStack.top();
```
use `pop_back_val` and remove instructions from the end in reverse-topo order?

kuhar commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,138 @@
+    return false;
+
+  // Invoke handling of "kill" instructions.
+  m_changed = false;
```
you could determine this based on `m_removalStack` being empty or not, right?

kuhar commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,138 @@
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
```
Can you describe how this function modifies the IR?

kuhar commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,138 @@
+  while (instIter != parentBlock->end()) {
+    if (m_instsForRemoval.insert(&*instIter).second)
+      m_removalStack.push(&*instIter);
+    instIter++;
```
nit: `++instIter`

kuhar commented on llpc/lower/llpcSpirvLowerTerminator.h:
```diff
@@ -0,0 +1,61 @@
+  SpirvLowerTerminator &operator=(const SpirvLowerTerminator &) = delete;
+
+  bool m_changed;
+  std::unordered_set<llvm::Instruction *> m_instsForRemoval; // Instructions to be removed
```
`DenseSet<llvm::Instruction*>`?

perlfu replied:
Use SmallPtrSet as the number of instructions is going to be small (typically zero).

kuhar commented on llpc/lower/llpcSpirvLowerTerminator.h:
```diff
@@ -0,0 +1,61 @@
+
+  bool m_changed;
+  std::unordered_set<llvm::Instruction *> m_instsForRemoval; // Instructions to be removed
+  std::stack<llvm::Instruction *> m_removalStack;
```
use `llvm::SmallVector`?

kuhar commented on llpc/test/shaderdb/OpKill_TestFunctionInline_lit.frag:
```diff
@@ -0,0 +1,37 @@
+    }
+}
+
+// BEGIN_SHADERTEST
```
Can you move this to the top of the file?

kuhar commented on llpc/test/shaderdb/OpKill_TestFunctionUnreachable_lit.spvasm:
```diff
@@ -0,0 +1,86 @@
+; RUN: amdllpc -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+
+; SHADERTEST-LABEL: {{^// LLPC.*}} SPIRV-to-LLVM translation results
+; SHADERTEST: unreachable
+
+; SHADERTEST-LABEL: {{^// LLPC.*}} SPIR-V lowering results
+; SHADERTEST: call void{{.*}} @lgc.create.kill
+; SHADERTEST-NEXT: br label
```
What does this test? There's very little context around the instructions being matched.

amdvlk-admin commented:
### Test summary for commit aa2a38f36e471c3a07cfc003de969844636a7567
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 0d602bbcfa1a86fd52f66966554b05e384a10d38</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1598ada40008302abedfd724275329f8e7dea637</li>
</ul></details>

<details><summary>CTS tests (Failed: 3/140859)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24005/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22954/46960 (48.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24004/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22955/46960 (48.9%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22646/46939 (48.2%)</li>
<li><details><summary>Failed: 1/46939 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 24292/46939 (51.8%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit f8d7001d3ca21c7026ad94ea495ad2f136caeb4a
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 0d602bbcfa1a86fd52f66966554b05e384a10d38</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1598ada40008302abedfd724275329f8e7dea637</li>
</ul></details>

<details><summary>CTS tests (Failed: 5/140859)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24005/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22954/46960 (48.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24004/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22955/46960 (48.9%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22644/46939 (48.2%)</li>
<li><details><summary>Failed: 3/46939 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail

FAILURE: dEQP-VK.renderpass.suballocation.formats.a2b10g10r10_uint_pack32.input.load.dont_care.self_dep_clear_draw
Stack trace: Flake

FAILURE: dEQP-VK.ubo.3_level_array.scalar.column_major_mat4x2.fragment_comp_access
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 24292/46939 (51.8%)</li>
</ul></details>


</ul></details>

nhaehnle reviewed (commented):
The overall change seems sound to me, but I do have a bunch of comments.

perlfu commented on llpc/lower/llpcSpirvLowerTerminator.h:
```diff
@@ -0,0 +1,61 @@
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  llpcSpirvLowerTerminator.h
+ * @brief LLPC header file: contains declaration of class Llpc::SpirvLowerTerminator.
+ ***********************************************************************************************************************
+ */
```
Sorry, I forgot we had decided to adopted LLVM common practice for classes we add/modify in this part of the compiler.

perlfu commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,138 @@
+  auto parentBlock = callInst.getParent();
+  auto instIter = std::next(callInst.getIterator());
+
+  // Already end of block?
+  if (instIter == parentBlock->end()) {
+    ReturnInst::Create(*m_context, parentBlock);
+    m_changed = true;
+    return;
+  }
```
Replace with assert.

amdvlk-admin commented:
### Test summary for commit b0dbc925d2b1d016661bf757a93c2d1029925bd9
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 0d602bbcfa1a86fd52f66966554b05e384a10d38</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1598ada40008302abedfd724275329f8e7dea637</li>
</ul></details>

<details><summary>CTS tests (Failed: 3/140859)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24005/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22954/46960 (48.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24004/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22955/46960 (48.9%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22646/46939 (48.2%)</li>
<li><details><summary>Failed: 1/46939 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 24292/46939 (51.8%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 1c0a07b4b5b21b6591f5e5f2f8ea81cf4203cfd6
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 0d602bbcfa1a86fd52f66966554b05e384a10d38</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1598ada40008302abedfd724275329f8e7dea637</li>
</ul></details>

<details><summary>CTS tests (Failed: 3/140859)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24005/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22954/46960 (48.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24004/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22955/46960 (48.9%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22646/46939 (48.2%)</li>
<li><details><summary>Failed: 1/46939 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 24292/46939 (51.8%)</li>
</ul></details>


</ul></details>

perlfu commented:
Pretty sure the CTS fails are not to do with this patch. I cannot reproduce them locally.

JaxLinAMD commented:
please ignore this failure: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
it's a known issue

nhaehnle commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,159 @@
+  // Invoke handling of "kill" instructions.
+  visit(m_module);
+
+  // Remove any dead instructions.
+  bool changed = !m_removalStack.empty();
+  while (!m_removalStack.empty()) {
+    auto deadInst = m_removalStack.pop_back_val();
+    LLVM_DEBUG(dbgs() << "remove: " << *deadInst << "\n");
+    deadInst->dropAllReferences();
+    deadInst->eraseFromParent();
+  }
+  m_instsForRemoval.clear();
```
Can we simplify this by only keeping m_instsForRemoval?

I know that it's a set, and we usually have to be careful about iteration order introducing non-determinism. But in this case we're only deleting the instructions, so...?

perlfu replied:
For the second test case I wrote (which is a made up theoretical example), it will fail if the instructions are not removed in reverse order as the second instruction is dependent on the first.  I just assume that was expected behaviour, but perhaps dropAllReference should be preventing that?

nhaehnle reviewed (approved):
Basically LGTM, but I wonder if it couldn't still be improved slightly?

JaxLinAMD commented:
@kuhar requested some changes,  not sure if i can merge now.

kuhar reviewed (approved):
Thanks for the fixes. LGTM.

JaxLinAMD commented:
retest this please

JaxLinAMD commented:
@perlfu , could you rebase your own branch on top of latest github/dev and then force-push?

amdvlk-admin commented:
### Test summary for commit 60cf01209d9496a1ba380492fd589aaecc377181
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev a87d46c4de072f579b9ea51faa6a5814c9a5bfb8</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 5cb4dc49f6f35597873e6255129b92d102aae22b</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

JaxLinAMD commented:
@kuhar LLPC Docker CI / Features: +clang+sanitizers, assertions: OFF this build takes too much time, are you able to access docker agents? 
LLPC Docker CI / Features: +clang+sanitizers, assertions: ON also there is no space left on this agent

perlfu commented:
This needs reworking slightly because there is an issue when the inliner runs on functions with return values. These already gain exit unification, which looks like unnecessary instructions to the pass as it stands.
Please hold off on merging @JaxLinAMD

JaxLinAMD commented:
got it

perlfu commented:
I have updated this to address the issue that came up in testing, and added an additional test to cover the problematic case.  I do not know if this needs reviewing again?

amdvlk-admin commented:
### Test summary for commit 6d2c9d9d60b9f35b54f6b81a60020307ced8e6ca
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 88613f67ca2e3e64cc1c7c4ccd47e126495e6e23</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

amdrexu commented:
>  I do not know if this needs reviewing again?

I think it's good to go. And I see this PR is ready for merging.

mariusz-sikora-at-amd commented on llpc/lower/llpcSpirvLowerTerminator.cpp:
```diff
@@ -0,0 +1,176 @@
+  if (!blockTerm)
+    return;
+  if (BranchInst *branch = dyn_cast<BranchInst>(blockTerm)) {
+    assert(branch->isUnconditional());
+    BasicBlock *targetBlock = branch->getSuccessor(0);
```
Why we are limiting this only to conditional branches and to successor(0)/IfTrue ?
Shouldn't we check both successors for conditional branch ? And if we have unconditional branch then successor based on the condition ?

I'm asking because I found by accident test which is crashing on the assertion L151.
angle_deqp_khr_gles31_tests --use-angle=vulkan --gtest_filter=KHR-GLES31.core.shader_image_load_store.basic-allFormats-store-fs
@perlfu @nhaehnle
